src/server/scripts/poller.js

```diff
@@ -0,0 +1,274 @@
+
+        if (!repo.get('store_name')) {
+          logger.info(`${owner}/${name}: NOT REGISTERED`);
+          return;
+        }
+
+        try {
+          if (await checker(owner, name, last_polled_at)) {
+            logger.info(`${owner}/${name}: NEEDSBUILD`);
+            await builder(owner, name);
+          } else {
+            logger.info(`${owner}/${name}: UNCHANGED`);
+          }
+        } catch (e) {
+          raven_client.captureException(e);
+          logger.error(`${owner}/${name}: FAILED (${e.message})`);
+        }
+        logger.info('==========');
+      });
+    });
+  });
+};
+
+
+// Whether a given snap (GitHub) repository has changed since 'last_polled_at'.
+// Consider changes in the repository itself as well as any of the (GitHub)
+// parts source.
+export const checkSnapRepository = async (owner, name, last_polled_at) => {
+  const token = conf.get('GITHUB_AUTH_CLIENT_TOKEN');
+  const repo_url = getGitHubRepoUrl(owner, name);
+  if (await hasRepoChanged(repo_url, last_polled_at, token)) {
+    logger.info(`The ${owner}/${name} repository has changed.`);
+    return true;
+  }
+  logger.info(`${owner}/${name}: unchanged, checking parts ...`);
+
+  let snapcraft_yaml;
+  try {
+    snapcraft_yaml = await internalGetSnapcraftYaml(owner, name, token);
+  } catch (e) {
+    return false;
+  }
+  for (const source_part of extractPartsToPoll(snapcraft_yaml.contents)) {
+    logger.info(`${owner}/${name}: Checking whether ${source_part.repoUrl} part has changed.`);
+    if (await source_part.hasRepoChangedSince(last_polled_at, token)) {
+      logger.info(`${owner}/${name}: ${source_part.repoUrl} changed.`);
+      return true;
+    }
+  }
+  return false;
+};
+
+
+// Request a build of a given snap repository (in LP) and update `polled_at` (in DB).
+export const buildSnapRepository = async (owner, name) => {
+  const _now = new Date().getTime();
+  // XXX request LP build
+  // const repo_url = getGitHubRepoUrl(owner, name);
+  return db.transaction(async (trx) => {
+    const row = await db.model('Repository')
+      .where({ owner, name })
+      .fetch({ transacting: trx });
+    await row.save({ polled_at: _now }, { method: 'update', transacting: trx });
+  });
+};
+
+
+// Whether a given (GitHub) repository has new commits since 'last_polled_at'.
+export const hasRepoChanged = async (repositoryUrl, last_polled_at, token) => {
+  if (last_polled_at === undefined || !last_polled_at) {
+    throw new Error('`last_polled_at` must be given.');
+  }
+  const last_polled = moment(last_polled_at);
+  const since = last_polled.toISOString();
+  const { owner, name } = parseGitHubRepoUrl(repositoryUrl);
+  const uri = `/repos/${owner}/${name}/commits?since=${since}`;
+  const options = {
+    token,
+    headers: {
+      'If-Modified-Since': last_polled.format('ddd, MM MMM YYYY HH:mm:ss [GMT]')
+    },
+    json: true
+  };
+
+  const response = await requestGitHub.get(uri, options);
+
+  switch (response.statusCode) {
```

Comment:
Worth handling 404? I've seen more than one deleted repo in https://launchpad.net/~build.snapcraft.io/+snaps.